cpp/src/neighbors/detail/hnsw.hpp

```diff
@@ -262,18 +266,392 @@ int initialize_point_in_hnsw(hnswlib::HierarchicalNSW<DistT>* appr_algo,
+  mmap_reader(const std::string& filename, size_t num_elements) : size_{num_elements * sizeof(T)}
+  {
+    fd_ = open(filename.c_str(), O_RDONLY);
+    if (fd_ == -1) { throw std::runtime_error("cuvs::mmap_reader error opening file"); }
+
+    float file_size_gb = size_ / 1e9;
+    RAFT_LOG_INFO("mmap opening file %s, num_elements [%zu] size %.2f GB",
+                  filename.c_str(),
+                  num_elements,
+                  file_size_gb);
+
+    data_ = mmap(nullptr, size_, PROT_READ, MAP_PRIVATE, fd_, 0);
+    // close(fd_);
+    if (data_ == MAP_FAILED) { throw std::runtime_error("cuvs::mmap_reader error"); }
+    if (madvise(data_, size_, MADV_HUGEPAGE) != 0) {
+      munmap(data_, size_);
+      data_ = nullptr;
+      throw std::runtime_error("cuvs::mmap_reader madvise error");
+    }
+  }
+
+  ~mmap_reader() noexcept
+  {
+    if (data_ != nullptr) { munmap(data_, size_); }
+    close(fd_);
+  }
+
+  mmap_reader(const mmap_reader& res)                      = delete;
+  auto operator=(const mmap_reader& other) -> mmap_reader& = delete;
+  // Moving is fine
+  mmap_reader(mmap_reader&& other)
+    : fd_{std::exchange(other.fd_, -1)},
+      data_{std::exchange(other.data_, nullptr)},
+      size_{std::exchange(other.size_, 0)}
+  {
+  }
+  auto operator=(mmap_reader&& other) -> mmap_reader&
+  {
+    std::swap(this->fd_, other.fd_);
+    std::swap(this->data_, other.data_);
+    std::swap(this->size_, other.size_);
+    return *this;
+  }
+
+  [[nodiscard]] auto view(idx_t n_rows, idx_t n_cols) const -> raft::host_matrix_view<T, idx_t>
+  {
+    return raft::make_host_matrix_view<T, idx_t>(
+      reinterpret_cast<T*>((char*)data_), n_rows, n_cols);
+  }
+  [[nodiscard]] auto view(idx_t n_rows) const -> raft::host_vector_view<T, idx_t>
+  {
+    return raft::make_host_vector_view<T, idx_t>(reinterpret_cast<T*>((char*)data_), n_rows);
+  }
+
+ private:
+  int fd_ = -1;
+  void* data_;
+  size_t size_;
+};
+
+struct buffered_ofstream {
+  buffered_ofstream(std::ostream* os, size_t buffer_size) : os_(os), buffer_(buffer_size), pos_(0)
+  {
+  }
+
+  ~buffered_ofstream() noexcept
+  {
+    // flush it
+    flush();
+  }
+
+  buffered_ofstream(const buffered_ofstream& res)                      = delete;
+  auto operator=(const buffered_ofstream& other) -> buffered_ofstream& = delete;
+  buffered_ofstream(buffered_ofstream&& other)                         = delete;
+  auto operator=(buffered_ofstream&& other) -> buffered_ofstream&      = delete;
+
+  void flush()
+  {
+    if (pos_ > 0) {
+      os_->write(reinterpret_cast<char*>(&buffer_.front()), pos_);
+      if (!os_->good()) { RAFT_FAIL("Error writing HNSW file!"); }
+      pos_ = 0;
+    }
+  }
+
+  void write(const char* input, size_t size)
+  {
+    if (pos_ + size > buffer_.size()) { flush(); }
+    std::copy(input, input + size, &buffer_[pos_]);
+    pos_ += size;
+  }
+
+ private:
+  std::vector<char> buffer_;
+  std::ostream* os_;
+  size_t pos_;
+};
+
```

Comment:
Please move these to file_io.hpp

Reply:
Done. I have removed `mmap_reader` and switched to batched `preads` instead.